node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+			}
+			goFields[i] = reflect.StructField{
+				Name: fieldNameFromSchema(field.Name()),
+				Type: ftyp,
+			}
+		}
+		return reflect.StructOf(goFields)
+	case *schema.TypeMap:
+		ktyp := inferGoType(typ.KeyType())
+		vtyp := inferGoType(typ.ValueType())
+		if typ.ValueIsNullable() {
+			vtyp = reflect.PtrTo(vtyp)
+		}
+		return reflect.MapOf(ktyp, vtyp)
+	case *schema.TypeList:
+		etyp := inferGoType(typ.ValueType())
+		if typ.ValueIsNullable() {
+			etyp = reflect.PtrTo(etyp)
+		}
+		return reflect.SliceOf(etyp)
+	case *schema.TypeUnion:
+		type goUnion struct {
+			Index int // 0..len(typ.Members)-1
+			Value interface{}
+		}
+		return reflect.TypeOf(goUnion{})
+	}
+	panic(fmt.Sprintf("%T\n", typ))
+}
+
+// Prototype implements an ipld.NodePrototype given a Go pointer type and an
+// IPLD schema type.
+//
+// In this form, it is assumed that the Go type and IPLD schema type are
+// compatible. TODO: check upfront and panic otherwise
+func Prototype(ptrType interface{}, schemaType schema.Type) ipld.NodePrototype {
+	goPtrType := reflect.TypeOf(ptrType)
+	if goPtrType.Kind() != reflect.Ptr {
+		panic("ptrType must be a pointer")
+	}
+	return prototype(goPtrType.Elem(), schemaType)
+}
+
+func prototype(goType reflect.Type, schemaType schema.Type) ipld.NodePrototype {
+	if goType.Kind() == reflect.Invalid {
+		panic("goType must be valid")
+	}
+	if schemaType == nil {
+		panic("schemaType must not be nil")
+	}
+	return &_prototype{schemaType: schemaType, goType: goType}
+}
+
+var (
+	_ ipld.NodePrototype = (*_prototype)(nil)
+
+	_ ipld.Node        = (*_node)(nil)
+	_ schema.TypedNode = (*_node)(nil)
+
+	_ ipld.NodeBuilder   = (*_builder)(nil)
+	_ ipld.NodeAssembler = (*_assembler)(nil)
+
+	_ ipld.MapAssembler = (*_structAssembler)(nil)
+	_ ipld.MapIterator  = (*_structIterator)(nil)
+
+	_ ipld.ListAssembler = (*_listAssembler)(nil)
+	_ ipld.ListIterator  = (*_listIterator)(nil)
+)
+
+type _prototype struct {
+	schemaType schema.Type
+	goType     reflect.Type // non-pointer
+}
+
+func (w *_prototype) NewBuilder() ipld.NodeBuilder {
+	return &_builder{_assembler{
+		schemaType: w.schemaType,
+		val:        reflect.New(w.goType).Elem(),
+	}}
+}
+
+// TODO: consider these Typed interfaces for the schema package
+
+type TypedPrototype interface {
+	ipld.NodePrototype
+
+	Representation() ipld.NodePrototype
+}
+
+type TypedAssembler interface {
+	ipld.NodeAssembler
+
+	Representation() ipld.NodeAssembler
+}
+
+func (w *_prototype) Representation() ipld.NodePrototype {
+	return (*_prototypeRepr)(w)
+}
+
+var (
+	goTypeString = reflect.TypeOf("")
+	goTypeBytes  = reflect.TypeOf([]byte{})
+	goTypeLink   = reflect.TypeOf((*ipld.Link)(nil)).Elem()
+
+	schemaTypeFieldName = schema.SpawnString("fieldNameString")
+)
+
+type _node struct {
+	schemaType schema.Type
+
+	val reflect.Value // non-pointer
+}
+
+// TODO: only expose TypedNode methods if the schema was explicit.
+// type _typedNode struct {
+// 	_node
+// }
+
+func (w *_node) Type() schema.Type {
```

Comment:
Because I initially called all these nodeWrapper, assemblerWrapper, etc. The names just got veeeery long and I didn't like them, so I went for _node, _assembler, etc.

I'm not too fond of `n _node` because `n` is usually `n int`, and `p _prototype` is usually `p []byte`, and `s _structAssembler` is usually `s string`... so I was a bit lazy and left that out for now, defaulting to `w` for all of them, which is almost never used in a standard way, and still gives the hint that these are "wrappers".